platforms/jvm/code-quality/src/main/java/org/gradle/api/plugins/quality/internal/CheckstyleInvoker.java

```diff
@@ -0,0 +1,279 @@
+    private static void assertUnsupportedReportFormatSARIF(VersionNumber version) {
+        throw new GradleException("SARIF report format is supported on Checkstyle versions 10.3.3 and newer. Please upgrade from Checkstyle " + version +" or disable the SARIF format.");
+    }
+
+    @Nullable
+    private static Node parseCheckstyleXml(Boolean isXmlRequired, File xmlOutputLocation) {
+        try {
+            return isXmlRequired ? new XmlParser().parse(xmlOutputLocation) : null;
+        } catch (IOException | SAXException | ParserConfigurationException e) {
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static String getMessage(Boolean isXmlRequired, File xmlOutputLocation, Boolean isHtmlRequired, File htmlOutputLocation, Boolean isSarifRequired, File sarifOutputLocation, Node reportXml) {
+        return String.format("Checkstyle rule violations were found.%s%s",
+            getReportUrlMessage(isXmlRequired, xmlOutputLocation, isHtmlRequired, htmlOutputLocation, isSarifRequired, sarifOutputLocation),
+            getViolationMessage(reportXml)
+        );
+    }
+
+    private static String getReportUrlMessage(Boolean isXmlRequired, File xmlOutputLocation, Boolean isHtmlRequired, File htmlOutputLocation, Boolean isSarifRequired, File sarifOutputLocation) {
+        File outputLocation;
+        if (isHtmlRequired) {
+            outputLocation = htmlOutputLocation;
+        } else if (isXmlRequired) {
+            outputLocation = xmlOutputLocation;
+        } else if (isSarifRequired) {
+            outputLocation = sarifOutputLocation;
+        } else {
+            outputLocation = null;
+        }
+        return outputLocation != null ? String.format(" See the report at: %s", new ConsoleRenderer().asClickableFileUrl(outputLocation)) : "\n";
+    }
+
+    private static String getViolationMessage(@Nullable Node reportXml) {
+        if (violationsExist(reportXml)) {
+            int errorFileCount = getErrorFileCount(reportXml);
+            List<String> violations = getViolations(reportXml);
+            return String.format("\n" +
+                "Checkstyle files with violations: %s\n" +
+                "Checkstyle violations by severity: %s\n",
+                errorFileCount,
+                violations
+            );
+        }
+        return "\n";
+    }
+
+    private static String readText(InputStream stream) {
+        try {
+            return IOUtils.toString(stream, StandardCharsets.UTF_8);
+        } catch (IOException e) {
+            throw new UncheckedIOException(e);
+        }
+    }
+
+    private static List<String> getViolations(Node reportXml) {
+        List<String> severity = new ArrayList<>();
+        NodeList errorNodes = reportXml.getAt(QName.valueOf("file")).getAt("error");
+        for (Object node : errorNodes) {
+            Node errorNode = (Node) node;
+            severity.add((String) errorNode.attribute("severity"));
+        }
+        return severity.stream()
```

Comment:
Can we keep this as stream-y as possible?
```suggestion
        @SuppressWarnings("unchecked")
        List<Node> errorNodes = reportXml.getAt(QName.valueOf("file")).getAt("error");
        return errorNodes.stream()
            .map(node -> (String) node.attribute("severity"))
```